pkg/kubelet/cm/devicemanager/manager.go

```diff
@@ -791,9 +791,36 @@ func (m *ManagerImpl) filterByAffinity(podUID, contName, resource string, availa
 		nodes = append(nodes, node)
 	}
 
-	// Sort the list of nodes by how many devices they contain.
+	// Sort the list of nodes by:
+	// 1) Nodes contained in the 'hint's affinity set
+	// 2) Nodes not contained in the 'hint's affinity set
+	// 3) The fake NUMANode of -1 (assuming it is included in the list)
+	// Within each of the groups above, sort the nodes by how many devices they contain
 	sort.Slice(nodes, func(i, j int) bool {
-		return perNodeDevices[i].Len() < perNodeDevices[j].Len()
+		// If one or the other of nodes[i] or nodes[j] is in the 'hint's affinity set
+		if hint.NUMANodeAffinity.IsSet(nodes[i]) && hint.NUMANodeAffinity.IsSet(nodes[j]) {
+			return perNodeDevices[nodes[i]].Len() < perNodeDevices[nodes[j]].Len()
+		}
+		if hint.NUMANodeAffinity.IsSet(nodes[i]) {
+			return true
+		}
+		if hint.NUMANodeAffinity.IsSet(nodes[j]) {
+			return false
+		}
+
+		// If either nodes[i] or nodes[j] are in the 'hint's affinity set (but are not -1)
+		if nodes[i] != nodeWithoutTopology && nodes[j] != nodeWithoutTopology {
+			return perNodeDevices[nodes[i]].Len() < perNodeDevices[nodes[j]].Len()
+		}
+		if nodes[i] != nodeWithoutTopology {
+			return true
+		}
+		if nodes[j] != nodeWithoutTopology {
+			return false
+		}
+
+		// If both nodes[i] and nodes[j] are -1
+		return perNodeDevices[nodes[i]].Len() < perNodeDevices[nodes[j]].Len()
```

Comment:
After seeing this written out, I actually think it would be clearer as:
```
// If one or the other of nodes[i] or nodes[j] is the fake NUMA node -1 (they can't both be)
if nodes[i] == nodeWithoutTopology {
	return false
}
if nodes[j] == nodeWithoutTopology {
	return true
}

// Otherwise both nodes[i] and nodes[j] are real NUMA nodes that are not in the 'hint's' affinity list.
return perNodeDevices[nodes[i]].Len() < perNodeDevices[nodes[j]].Len()
```

Sorry for all the churn.

Reply:
Updated PTAL

Reply:
I don't see it updated to this.

Reply:
I see now. You added this, but didn't remove the code it was supposed to replace.